File: server/src/schemas/ml.py
"""ML prediction schemas."""

from pydantic import BaseModel


class StudentFeatures(BaseModel):
    """Feature vector used for prediction."""

    average_grade: float
    min_grade: float
    max_grade: float
    grade_std: float
    grade_trend_slope: float
    num_subjects: int
    failing_subjects: int
    absence: int
    absence_justified: int
    late: int
    disturbance: int
    total_absences: int
    total_negative_events: int
    total_positive_events: int


class StudentPrediction(BaseModel):
    """Prediction result for a single student."""

    student_tz: str
    student_name: str
    predicted_grade: float
    dropout_risk: float
    risk_level: str
    features: StudentFeatures


class TrainResponse(BaseModel):
    """Training result with metrics."""

    status: str
    samples: int
    grade_model_mae: float
    grade_model_median_ae: float | None = None
    dropout_model_accuracy: float
    dropout_model_roc_auc: float | None = None
    dropout_model_pr_auc: float | None = None
    dropout_precision_high_risk: float | None = None
    dropout_recall_high_risk: float | None = None
    dropout_confusion_matrix: list[list[int]] | None = None
    grade_feature_importances: dict[str, float]
    dropout_feature_importances: dict[str, float]
    class_distribution: dict[str, int] | None = None
    cv_folds_grade: int | None = None
    cv_folds_dropout: int | None = None
    evaluation_strategy: str | None = None
    high_risk_threshold: float | None = None
    medium_risk_threshold: float | None = None


class BatchPredictionResponse(BaseModel):
    """Batch prediction response."""

    predictions: list[StudentPrediction]
    model_trained: bool
    total_students: int
    total: int
    page: int
    page_size: int
    high_risk_count: int
    medium_risk_count: int


class ModelStatusResponse(BaseModel):
    """Model status information."""

    trained: bool
    trained_at: str | None = None
    samples: int | None = None
    grade_model_mae: float | None = None
    grade_model_median_ae: float | None = None
    dropout_model_accuracy: float | None = None
    dropout_model_roc_auc: float | None = None
    dropout_model_pr_auc: float | None = None
    dropout_precision_high_risk: float | None = None
    dropout_recall_high_risk: float | None = None
    evaluation_strategy: str | None = None
    cv_folds_grade: int | None = None
    cv_folds_dropout: int | None = None

File: client/src/lib/types/ml.ts
import { z } from "zod";

export const TrainResponseSchema = z.object({
    status: z.string(),
    samples: z.number(),
    grade_model_mae: z.number(),
    grade_model_median_ae: z.number().nullable().optional(),
    dropout_model_accuracy: z.number(),
    dropout_model_roc_auc: z.number().nullable().optional(),
    dropout_model_pr_auc: z.number().nullable().optional(),
    dropout_precision_high_risk: z.number().nullable().optional(),
    dropout_recall_high_risk: z.number().nullable().optional(),
    dropout_confusion_matrix: z.array(z.array(z.number())).nullable().optional(),
    grade_feature_importances: z.record(z.string(), z.number()),
    dropout_feature_importances: z.record(z.string(), z.number()),
    class_distribution: z.record(z.string(), z.number()).nullable().optional(),
    cv_folds_grade: z.number().nullable().optional(),
    cv_folds_dropout: z.number().nullable().optional(),
    evaluation_strategy: z.string().nullable().optional(),
    high_risk_threshold: z.number().nullable().optional(),
    medium_risk_threshold: z.number().nullable().optional(),
});
export type TrainResponse = z.infer<typeof TrainResponseSchema>;

export const StudentFeaturesSchema = z.object({
    average_grade: z.number(),
    min_grade: z.number(),
    max_grade: z.number(),
    grade_std: z.number(),
    grade_trend_slope: z.number(),
    num_subjects: z.number(),
    failing_subjects: z.number(),
    absence: z.number(),
    absence_justified: z.number(),
    late: z.number(),
    disturbance: z.number(),
    total_absences: z.number(),
    total_negative_events: z.number(),
    total_positive_events: z.number(),
});

export const StudentPredictionSchema = z.object({
    student_tz: z.string(),
    student_name: z.string(),
    predicted_grade: z.number(),
    dropout_risk: z.number(),
    risk_level: z.enum(["low", "medium", "high"]),
    features: StudentFeaturesSchema,
    factors: z.array(z.string()).optional(),
});
export type StudentPrediction = z.infer<typeof StudentPredictionSchema>;

export const BatchPredictionResponseSchema = z.object({
    predictions: z.array(StudentPredictionSchema),
    model_trained: z.boolean().optional(),
    total_students: z.number().optional(),
    total: z.number(),
    page: z.number(),
    page_size: z.number(),
    high_risk_count: z.number(),
    medium_risk_count: z.number(),
});
export type BatchPredictionResponse = z.infer<typeof BatchPredictionResponseSchema>;

export const ModelStatusResponseSchema = z.object({
    trained: z.boolean(),
    trained_at: z.string().nullable(),
    samples: z.number().nullable(),
    grade_model_mae: z.number().nullable(),
    grade_model_median_ae: z.number().nullable().optional(),
    dropout_model_accuracy: z.number().nullable(),
    dropout_model_roc_auc: z.number().nullable().optional(),
    dropout_model_pr_auc: z.number().nullable().optional(),
    dropout_precision_high_risk: z.number().nullable().optional(),
    dropout_recall_high_risk: z.number().nullable().optional(),
    evaluation_strategy: z.string().nullable().optional(),
    cv_folds_grade: z.number().nullable().optional(),
    cv_folds_dropout: z.number().nullable().optional(),
});
export type ModelStatusResponse = z.infer<typeof ModelStatusResponseSchema>;
